### PR TITLE
finished database week2 homework

### PR DESCRIPTION
**Pull request by chunmeishui**

### CODE REVIEW

File: databases/week2/School database.sql [hunk 1]
```diff
@@ -0,0 +1,26 @@
+CREATE DATABASE school DEFAULT CHARACTER SET = 'utf8mb4';
+USE school;
+-- Class: with the columns: id, name, begins (date), ends (date)
+CREATE TABLE `Class` (
+  `id` int(10) unsigned NOT NULL AUTO_INCREMENT PRIMARY KEY,
+  `name` varchar(255) NOT NULL,
+  `begin_date` DATETIME NOT NULL,
```
Comment (mani-zaeim) [inline databases/week2/School database.sql:7]: suppose to be begins and ends

File: databases/week2/School database.sql [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+CREATE TABLE `Student` (
+  `id` int(10) unsigned NOT NULL AUTO_INCREMENT PRIMARY KEY,
+  `name` varchar(255) NOT NULL,
+  `email` varchar(255) NOT NULL,
+  `phone` varchar(255) NULL,
+  `class_id` INT(10) UNSIGNED,
+  FOREIGN KEY (`class_id`) REFERENCES `Class` (`id`) ON DELETE CASCADE ON UPDATE CASCADE
+) ENGINE = InnoDB DEFAULT CHARSET = utf8mb4 COLLATE = utf8mb4_unicode_ci;
+-- Add a new column to the class table named status which can only have the following values: not-started, ongoing, finished (hint: enumerations).
+
```
Comment (mani-zaeim) [inline databases/week2/School database.sql:20]: the index and enum still missing, the index can be something like:
ALTER TABLE `student` ADD FULLTEXT INDEX `idx_school_student_name` (`name`);
  Hi mani; can i use  
   CREATE INDEX idx_school_student_name
 ON Student (name);?
 thanks for telling.

File: databases/week2/School_query.sql [hunk 3]
```diff
@@ -0,0 +1,85 @@
+
+SELECT * FROM Class;
+ALTER TABLE
+  Class
+ADD
+  status varchar(255);
```
Comment (mani-zaeim) [inline databases/week2/School_query.sql]: the enum is missing

File: databases/week2/hyf_lesson2_homework.sql [hunk 4]
```diff
@@ -0,0 +1,93 @@
+-- Part 1: Working with tasks
+INSERT INTO
+  task(
+    title,
+    description,
+    created,
+    updated,
+    due_date,
+    status_id
```
Comment (mani-zaeim) [inline databases/week2/hyf_lesson2_homework.sql:9]: the user_id column is missing

File: databases/week2/hyf_lesson2_homework.sql [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+UPDATE
+  task
+SET
+  status_id = 2
+WHERE
+  id = 36;
+UPDATE
+  task
+SET
+  status_id = 3
```
Comment (mani-zaeim) [inline databases/week2/hyf_lesson2_homework.sql:41]: its correct, but, is it possible to do this in another way? lets say I know there is a status called 'Done', is it possible to update it to 'Done' without knowing what is the id of the corresponding status in this case 'Done' ?

File: databases/week2/School database.sql (hunk 2 above)
Comment (chunmeishui) [reply databases/week2/School database.sql:20]: Hi, Mani, Marta said we do not have to do that part.so I did not do it.

File: databases/week2/School_query.sql (hunk 3 above)
Comment (chunmeishui) [reply databases/week2/School_query.sql]: when we add a column we do not have to write the column, but when we modify or drop we need to write the column.

File: databases/week2/hyf_lesson2_homework.sql [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+SELECT
+  task.title,
+  user.email
+FROM
+  task
+  JOIN user_task ON task.id = user_task.task_id
+  JOIN user ON user.id = user_task.user_id
+WHERE
+  user.email LIKE "%@spotify.com";
+SELECT
```
Comment (mani-zaeim) [inline databases/week2/hyf_lesson2_homework.sql:58]: its nice to assign the alias, just let you know its not necessary, I know if you dont the VSCode extension will show wrong result, but thats just the VSCode extension, if you use other IDE, it shows correct result

File: databases/week2/hyf_lesson2_homework.sql [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+  task
+  JOIN status ON status.id = task.status_id
+  JOIN user_task ON task.id = user_task.task_id
+  JOIN user ON user.id = user_task.user_id
+WHERE
+  user.name = "Donald Duck"
+  AND status.name = "Not started";
+SELECT
+  user.name,
+  COUNT(task.id)
```
Comment (mani-zaeim) [inline databases/week2/hyf_lesson2_homework.sql:72]: the count is not requested, however its a good idea to add Task Title, and Task Created date or Created Month.

File: databases/week2/hyf_lesson2_homework.sql [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+FROM
+  task
+  JOIN user_task ON task.id = user_task.task_id
+  JOIN user ON user.id = user_task.user_id
+WHERE
+  user.name = "Maryrose Meadows"
+GROUP BY
+  month(task.created) = 9;
+SELECT
+  count(task.id)
```
Comment (mani-zaeim) [inline databases/week2/hyf_lesson2_homework.sql]: the query is missing a column, how should we know the counts are belonging to which month ? please add column into selected fields inorder to be more informative

File: databases/week2/hyf_lesson2_homework.sql [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+FROM
+  task
+GROUP BY
+  month(created);
+-- changed the created of id = 35 to month 11th .
+UPDATE
```
Comment (mani-zaeim) [inline databases/week2/hyf_lesson2_homework.sql:88]: I dont know why did you add this query, but its not part of assignment, in result it can have negative point as it is not requested :)

File: databases/week2/library-database.sql [hunk 10]
```diff
@@ -0,0 +1,53 @@
+use library;
```
Comment (mani-zaeim) [inline databases/week2/library-database.sql:1]: i am sure you created the database/schema, however its always a good idea to add the create statement as well as use statement

File: databases/week2/library-database.sql [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+    PRIMARY KEY(`book_id`, `author_id`),
+    CONSTRAINT `fk_book_book_author` FOREIGN KEY (`book_id`) REFERENCES `book` (`id`) ON DELETE CASCADE ON UPDATE CASCADE,
+    CONSTRAINT `fk_author_book_author` FOREIGN KEY (`author_id`) REFERENCES `author` (`id`) ON DELETE CASCADE ON UPDATE CASCADE
+  ) ENGINE = InnoDB DEFAULT CHARSET = utf8mb4 COLLATE = utf8mb4_unicode_ci;
+  CREATE TABLE `borrowLog` (
+  `id` int(10) unsigned NOT NULL AUTO_INCREMENT PRIMARY KEY,
+  `borrow_date` DATETIME NOT NULL,
+  `return_date` DATETIME NOT NULL
+) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci;
+CREATE TABLE `book_member_borrowLog` (
```
Comment (mani-zaeim) [inline databases/week2/library-database.sql:45]: how about if we another member borrow the same book in same borrow date period?

File: databases/week2/library-database.sql [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+    `id` int(10) unsigned NOT NULL AUTO_INCREMENT PRIMARY KEY,
+    `name` varchar(255) NOT NULL,
+    `phone` varchar(255) NULL,
+    `email` varchar(255) NOT NULL
+  ) ENGINE = InnoDB DEFAULT CHARSET = utf8mb4 COLLATE = utf8mb4_unicode_ci;
+CREATE TABLE `book` (
+    `id` int(10) unsigned NOT NULL AUTO_INCREMENT PRIMARY KEY,
+    `name` varchar(255) NOT NULL,
+    `description` text NULL DEFAULT NULL,
+    `price` DATETIME NOT NULL,
```
Comment (mani-zaeim) [inline databases/week2/library-database.sql:27]: you need to change the Datatype of the price

File: databases/week2/library-query.sql [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+  member(name, phone, email)
+VALUES
+  ("sunny", "55668963", "sunny@163.com"),("ammy", "23368963", "sunny@163.com"),("jone", "48958963", "jone@163.com"),("lina", "45268963", "lina@163.com"),("fie", "25768963", "fie@163.com");
+-- rack
+INSERT INTO
+  rack(categary)
+VALUES
+  ("a-d"),("e-i"),("j-n"),("o-t"),("u-z");
+-- book
+ALTER TABLE
```
Comment (mani-zaeim) [inline databases/week2/library-query.sql:34]: good job inorder to understand the ALTER table and Modify option

File: databases/week2/library-query.sql [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+  book
+MODIFY
+  COLUMN price INT NOT NULL;
+UPDATE
+  book
+SET
+  price = 199
+WHERE
+  id = 9;
+-- change the price to varchar type
```
Comment (mani-zaeim) [inline databases/week2/library-query.sql:51]: as you commented, if you want to add the currency, you have to change it into varchar or use money datatype, however as the MySQL doesnt support Money, you may change it to Decimal, Flout or Double and ass a Column, call Currency

File: databases/week2/library-query.sql [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+INSERT INTO
```
Comment (mani-zaeim) [inline databases/week2/library-query.sql:52]: so because the price is not varchar, this insert statement will fail

File: databases/week2/library-query.sql [hunk 16]
```diff
@@ -0,0 +1,147 @@
...
+    '2017-10-25 15:05:09'
+  ),(
+    '2017-10-17 09:54:16',
+    '2017-10-27 13:05:09'
+  ),(
+    '2017-10-17 10:54:16',
+    '2017-10-27 14:05:09'
+  );
+-- book_author
+INSERT INTO
```
Comment (mani-zaeim) [inline databases/week2/library-query.sql:110]: some Author need to inserted before this insert statement :) so I will stop my review until here


Hi mani, I did insert the author in the first one. what do you mean here? would you pls tell me? Thanks.

**Review by mani-zaeim — COMMENTED** [review]
Good job, please add/correct the missing parts

File: databases/week2/School database.sql (hunk 2 above)
Comment (mani-zaeim) [reply databases/week2/School database.sql:20]: I know, i just added in comment so you can take a look if you have time :) otherwise thats not an issue

File: databases/week2/School_query.sql (hunk 3 above)
Comment (martamatos) [reply databases/week2/School_query.sql]: the type of the status table should be an enumeration not a varchar, the idea si to make you students google a bit ;p

File: databases/week2/library-database.sql [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+    CONSTRAINT `fk_rack` FOREIGN KEY (`rack_id`) REFERENCES `rack` (`id`) ON DELETE CASCADE ON UPDATE CASCADE
+  ) ENGINE = InnoDB DEFAULT CHARSET = utf8mb4 COLLATE = utf8mb4_unicode_ci;
+CREATE TABLE `book_author` (
+    `book_id` int(10) unsigned NOT NULL,
+    `author_id` int(10) unsigned NOT NULL,
+    PRIMARY KEY(`book_id`, `author_id`),
+    CONSTRAINT `fk_book_book_author` FOREIGN KEY (`book_id`) REFERENCES `book` (`id`) ON DELETE CASCADE ON UPDATE CASCADE,
+    CONSTRAINT `fk_author_book_author` FOREIGN KEY (`author_id`) REFERENCES `author` (`id`) ON DELETE CASCADE ON UPDATE CASCADE
+  ) ENGINE = InnoDB DEFAULT CHARSET = utf8mb4 COLLATE = utf8mb4_unicode_ci;
+  CREATE TABLE `borrowLog` (
```
Comment (martamatos) [inline databases/week2/library-database.sql:40]: this table is ok, but it could also be merged with the book_member_borrowLog table
basically you'd end up with a "borrow" table with the columns "id, book_id, member_id, borrow_date, return_date". if you do this you probably don't want to make (book_id, member_id) your primary key though, because then the same person wouldn't be able to borrow the same book twice.
yes, Marta.the same person can not borrow the same book twice and each time only one person can borrow the books. because the time record is up to seconds.
Thanks
